File: PACE/Tools/src/minfill/data/EmptySet.java
package minfill.data;

import java.util.Iterator;

public class EmptySet<T> implements Set<T> {
    private final static EmptySetIterator iterator = new EmptySetIterator();
    private static final EmptySet instance = new EmptySet();

    @SuppressWarnings("unchecked")
    public static <T> Set<T> instance() {
        return instance;
    }

    private EmptySet() {}

    @Override
    public boolean isEmpty() {
        return true;
    }

    @Override
    public boolean isProperSubsetOf(Set<T> other) {
        return !other.isEmpty();
    }

    @Override
    public boolean isSubsetOf(Set<T> other) {
        return true;
    }

    @Override
    public boolean contains(T element) {
        return false;
    }

    @Override
    public int size() {
        return 0;
    }

    @Override
    public Set<T> add(T element) {
        return new ImmutableSet<>(element);
    }

    @Override
    public Set<T> union(Set<T> other) {
        return other;
    }

    @Override
    public Set<T> intersect(Set<T> other) {
        return this;
    }

    @Override
    public Set<T> minus(Set<T> other) {
        return this;
    }

    @Override
    public Set<Set<T>> subsetsOfSizeAtMost(int size) {
        return new EmptySet<>();
    }

    @Override
    @SuppressWarnings("unchecked")
    public Iterator<T> iterator() {
        return iterator;
    }

    @Override
    public String toString() {
        return "[]";
    }

    static class EmptySetIterator<T> implements Iterator<T> {
        @Override
        public boolean hasNext() {
            return false;
        }

        @Override
        public T next() {
            throw new IllegalStateException("Empty iterator");
        }
    }
}

File: PACE/cppTools/src/minfill/sets/EmptySet.cpp
#include<functional>
#include "Set.h"
#include<iostream>

namespace minfill::sets {
	template <class T>
	class EmptySet : public Set<T>
	{
		private:
			constexpr EmptySet() {}
			constexpr static const EmptySet<T>& instance = EmptySet<T>();
		public:
			bool IsEmpty() const { return true; }
			bool IsProperSubsetOf(const Set<T>& other) const { return !other.IsEmpty(); }
			bool IsSubsetOf(const Set<T>& other) const { return true; }
			bool Contains(const T& elem) const { return false; }
			int Size() const { return 0; }
			const Set<T>& Add(const T& elem) const { return *this; } // TODO
			const Set<T>& Remove(const T& elem) const { return *this; }
			const Set<T>& Union(const Set<T>& other) const { return other; }
			const Set<T>& Intersect(const Set<T>& other) const { return *this; }
			const Set<T>& Minus(const Set<T>& other) const { return *this; }
			static const Set<T>& GetInstance() { return instance; }
	};
}

using namespace minfill::sets;
using namespace std;

int main(int argc, char* argv[]) {
	const Set<int>& set = EmptySet<int>::GetInstance();

	cout << set.Size() << endl;
}
